### PR TITLE
Issue #6649: add InvalidJavadocPositionCheck

### PR DESCRIPTION
**Pull request by rnveach**
Issue #6649
Contribution: https://github.com/checkstyle/contribution/pull/383

Regression to come.

### CODE REVIEW

**Comment by rnveach** [comment]
Regression: http://rveach.no-ip.org/checkstyle/regression/reports/247/

I used this check for a previous PR and didn't find anymore false positives.

**Comment by rnveach** [comment]
@pbludov Feel free to  review too.

File: src/xdocs/config_javadoc.xml [hunk 1]
```diff
@@ -154,6 +154,58 @@
       </subsection>
     </section>
 
+    <section name="InvalidJavadocPosition">
+      <p>Since Checkstyle 8.21</p>
+      <subsection name="Description" id="InvalidJavadocPosition_Description">
+        <p>
+          Checks that Javadocs are located at the correct position.
```
Comment (pbludov) [inline src/xdocs/config_javadoc.xml]: Please add the link to the specification:
https://docs.oracle.com/en/java/javase/11/docs/specs/doc-comment-spec.html

**Comment by pbludov** [comment]
> Regression: http://rveach.no-ip.org/checkstyle/regression/reports/247/

An excellent example of how to lose the doc comment:
http://rveach.no-ip.org/checkstyle/regression/reports/247/spotbugs/xref/spotbugs/src/gui/edu/umd/cs/findbugs/gui2/AnalyzingDialog.java.html#L44

@rnveach please add rationale and an example like this
```
@SuppressWarnings("serial")
/**
 * This comment looks like javadoc but has an invalid location.
 * Therefore, the text will not get into TestClass.html.
 */
public class TestClass {
}
```

**Comment by pbludov** [comment]
Inspired by:
http://rveach.no-ip.org/checkstyle/regression/reports/247/Orekit/xref/src/main/java/org/orekit/files/sp3/SP3Parser.java.html#L514

Foreign comment:
```
public class TestClass {
 /** The documentation for FOO. */
 // public static int FOO = 1;
 
 public static int BAR = 2;
}
```
TestClass.html will contain

Modifier and Type | Field | Description
-- | -- | --
static int | BAR | The documentation for FOO.

This can not be detected by this check. But it could be dangerous. Perhaps in the future we need an option for such cases.

File: src/xdocs/config_javadoc.xml (hunk 1 above)
Comment (rnveach) [reply src/xdocs/config_javadoc.xml]: Done.

**Comment by rnveach** [comment]
> Perhaps in the future we need an option for such cases.

Only if there is a way to identify commented code. There was another issue for a check like that but there is no easy way to identify if a comment is code or just an example or just text.

**Comment by rnveach** [comment]
> please add rationale and an example like this

Done.

**Comment by romani** [comment]
@pbludov , please finish review. I will be second reviewer.

**Comment by romani** [comment]
@rnveach , please make spell checker happy.

**Comment by romani** [comment]
Code looks good.

> Regression: http://rveach.no-ip.org/checkstyle/regression/reports/247/

please recheck status of server, this link is not reachable. 

-------

should we use it google_style.xml ?

we have big section about javadoc
https://checkstyle.org/styleguides/google-java-style-20170228.html#s7-javadoc
it kind of imply location to be appropriate for javadoc tool. I thinks we should place it.
we can put it at https://checkstyle.org/google_style.html#a7.1.1 .

----

should we put this Check to sun_style.xml ?
we never did this before .... 
https://checkstyle.org/styleguides/sun-code-conventions-19990420/CodeConventions.doc4.html#a16838
but @Vampire , moved us to next level, and we can step by step improve coverage of sun_style.xml

@pbludov , please share your opinion on usage of new Check in google and sun styles.

**Comment by rnveach** [comment]
@romani 
> this link is not reachable.

Try again. I didn't start my VM at the time.

> should we use it google_style.xml ?

I think it is a stretch that anything in the google style refers to correct javadoc placement. 

> should we put this Check to sun_style.xml ?

I agree it should be added to sun. It makes clear notes of good/bad positions.

**Comment by rnveach** [comment]
> sun_style.xml

There are some bugs with the sun style test area. I will provide some fixes for them.

**Comment by romani** [comment]
> I think it is a stretch that anything in the google style refers to correct javadoc placement.

please rephrase, am not sure I understand this clearly.

> @pbludov , please share your opinion on usage of new Check in google and sun styles.

@pbludov , please do.

**Comment by romani** [comment]
> Try again. 

results are impressive, good Check.

Concerns:
1) http://rveach.no-ip.org/checkstyle/regression/reports/247/openjdk10/index.html#A56
exception ?
there are 217 "Got an exception". Please update config to skip such files.

2)
http://rveach.no-ip.org/checkstyle/regression/reports/247/RxJava/index.html#A10 (good javadoc comment, right after header(unfortunately header is javadoc styled)). Is there way to reconcile this ? can we filter/suppress such cases ? Even openJdk use header in form of javadoc - http://rveach.no-ip.org/checkstyle/regression/reports/247/openjdk10/xref/src/java.base/share/classes/java/lang/Thread.java.html#L1 , and give whole world example :) on how to do this. It is too late fix this. We need to account this.

3) http://rveach.no-ip.org/checkstyle/regression/reports/247/openjdk10/index.html#A1292
kind of misusage of javadocs style comment, but looks like massive. Can we provide some sort of suppression on such/similar cases ?
Looks like only chekcstyle team consider type of multiline comment seriously in the world, all other even in openJdk , use whatever comment if it is NOT picked up by javadoc tool. We need to be prepared that in some cases or by some patterns users will keep comments them as is.

**Comment by rnveach** [comment]
The correct way to reconcile these is to change the javadoc to a comment. The check is correct that this isn't the place for a javadoc comment and it will be ignored by the javadoc tool. There serves no benefit to leave them as javadocs and not just convert them to comments.

> can we filter/suppress such cases ?

Yes, anyone can suppress them with our suppression files and xpath. Do we really need something built into the check for this? If so, are you suggesting we ignore all javadocs over package and imports?

**Comment by rnveach** [comment]
> I think it is a stretch that anything in the google style refers to correct javadoc placement.

I don't feel the google style in any language I saw refer to correct javadoc placement. My assumption is it just refers to how they should be styled and assumed the users would know the correct spot for them to be placed in.

**Comment by rnveach** [comment]
> "Got an exception". Please update config to skip such files.

http://rveach.no-ip.org/checkstyle/regression/reports/249/

**Comment by romani** [comment]
>Do we really need something built into the check for this? 

No, just curious how we ( user in future) can suppress such case. Example of suppression might be good for xdoc.

------

I more worries about point "2.".  What kind of suppression we would propose ? Too much violations, users most likely disable Check. I do not say such code is good, but users always looks jdk for examples how to do something.

**Comment by pbludov** [comment]
> @pbludov , please share your opinion on usage of new Check in google and sun styles.
> should we put this Check to sun_style.xml ?
Yes. Sun code conventions strictly define where doc comments should be and where not.
 
> should we use it google_style.xml ?
I vote to add this check to google_style.xml. The Google Java style guide does not explicitly prohibit doc comments in the wrong place, but it strictly limits the scope of comments to classes, interfaces, and methods.

**Comment by romani** [comment]
@pbludov , it is always better to quote a line from style that your reference to

My reading of style:

1)
https://checkstyle.org/styleguides/google-java-style-20180523/javaguide.html#s7.3-javadoc-where-required

> At the minimum, Javadoc is present for every public class, and every public or protected member of such a class, with a few exceptions noted below.

This can be used as requirements to place such javadoc on proper place. So I think Check need to be activated.  

2)
https://checkstyle.org/styleguides/google-java-style-20180523/javaguide.html#s7.3.4-javadoc-non-required

> Whenever an implementation comment would be used to define the overall purpose or behavior of a class or member, that comment is written as Javadoc instead (using /**).
 
@rnveach , please help us read this statement properly. Is it a demand to place details to javadoc ? So it could be interpreted as no `/** */` inside a method bod. So it mean is comment is in javadoc style it should be on proper place.

-------

@rnveach , please review our comments and think one more type on activation in google style. Looks like I and @pbludov are for activation.

**Comment by rnveach** [comment]
> please help us read this statement properly. Is it a demand to place details to javadoc ?
> Whenever an implementation comment would be used to define the overall purpose or behavior of a class or member, that comment is written as Javadoc instead (using /**).

implementation comment = saying how the class is implemented, like reading the code at a high level and in plain english.

It just says when the comment describes the implementation, the comment that is used should be a Javadoc. 

So it demands that any implementation comment can't be anything other than a Javadoc. It doesn't say where the javadoc should be located. One could argue it is asking for Javadoc comments inside a method since sometimes people write implementation comments for sections of code.
Guava only had 1 example, and I am not sure I would call it an implementation comment. http://rveach.no-ip.org/checkstyle/regression/reports/247/guava/xref/guava-tests/test/com/google/common/collect/MutableClassToInstanceMapTest.java.html#L92
I don't know any other google projects.

>  Looks like I and @pbludov are for activation.

I will add it then.
So add it to https://checkstyle.org/google_style.html#a7.1.1  ?

**Comment by romani** [comment]
> So add it to https://checkstyle.org/google_style.html#a7.1.1 ?

I would rather place it to https://checkstyle.org/styleguides/google-java-style-20180523/javaguide.html#s7.3.4-javadoc-non-required

to cover:
> Whenever an implementation comment would be used to define the overall purpose or behavior of a class or member, that comment is written as Javadoc instead (using /**).

As your Check is detecting ` instead (using /**).` - misplaced multiline comment that have javadoc identificator. True to say if you and me read style differently, other users might do the same ..... We can place in both section: 7.1.1 and 7.3.4 :) to cover most of readers.

**Comment by romani** [comment]
@rnveach , is there anything else that blocks you?s

**Comment by rnveach** [comment]
Just free time.

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/invalidjavadocposition/InputInvalidJavadocPosition.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    <T> T method3() { return null; }
+    /** valid */
+    String[] method4() { return null; }
+
+    void /** violation */ method20() {}
+    void method21 /** violation */ () {}
+    void method22(/** violation */) {}
+    void method23() /** violation */ {}
+    void method24() { /** violation */ }
+    void method25() { /** violation */ int variable; }
```
Comment (rnveach) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/invalidjavadocposition/InputInvalidJavadocPosition.java]: Found by IT's warn double checker.
A violation isn't being produced here. It thinks the variable is a field.

**Comment by rnveach** [comment]
I pushed an update with everything we talked about.

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/invalidjavadocposition/InputInvalidJavadocPosition.java (hunk 2 above)
Comment (rnveach) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/invalidjavadocposition/InputInvalidJavadocPosition.java]: This has not been fixed. It is outside the code of this PR, so it will have to be another issue/PR.

**Comment by rnveach** [comment]
Rebased to fix CI.

File: src/main/resources/sun_checks.xml [hunk 3]
```diff
@@ -87,6 +87,7 @@
 
         <!-- Checks for Javadoc comments.                     -->
         <!-- See http://checkstyle.sourceforge.net/config_javadoc.html -->
+        <module name="InvalidJavadocPosition"/>
         <module name="JavadocMethod"/>
         <module name="JavadocType"/>
```
Comment (romani) [inline src/main/resources/sun_checks.xml]: Should we name Check "JavadocPosition" ?

File: src/xdocs/checks.xml [hunk 4]
```diff
@@ -364,6 +364,12 @@
           Checks that interface type parameter names conform to a format specified
         by the format property.</td>
       </tr>
+      <tr>
+        <td><a href="config_naming.html#InvalidJavadocPosition">
+          InvalidJavadocPosition</a></td>
+        <td>
+          Checks that Javadocs are located at the correct position.</td>
```
Comment (romani) [inline src/xdocs/checks.xml]: this summary kind of prove that we need to drop "Invalid" from name of Check.

File: src/main/resources/sun_checks.xml (hunk 3 above)
Comment (romani) [reply src/main/resources/sun_checks.xml]: @pbludov , please share your ideas on this ?
https://checkstyle.org/config_naming.html , we can put "Invalid" to almost all Checks, but we do not do this.
It would be better for us to make some concept of module naming.
We have now that much [Checks with prefixes](https://checkstyle.org/checks.html): "Avoid" , "Illegal", "Missing", "No", "Simplify".

In my mind, javadoc position is very certain, it could not several acceptable positions and some position are invalid. So we just check position on certain place, so "JavadocPosition".

**Comment by romani** [comment]
@rnveach , please address minors while name is in discussion.

File: src/main/resources/sun_checks.xml (hunk 3 above)
Comment (romani) [reply src/main/resources/sun_checks.xml]: to not block this awesome PR that much, lets vote one more time on name.

Comment (pbludov) [reply src/main/resources/sun_checks.xml]: > @pbludov , please share your ideas on this ?

It is a good point. We don't have checks with the prefix `Invalid`. But simple `JavadocPosition` may be not clear enough. By the way, this check is not about *positions*. It is about *relations*. Each javadoc comment should be bound to some code. If there is no connection, it is a violation. That's what this check does. We need to show it in the name of the check somehow.

Comment (romani) [reply src/main/resources/sun_checks.xml]: as `JavadocPosition` is not clear also, and you have no better name, lets keep `InvalidJavadocPosition`.
We can not keep PR that long.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/InvalidJavadocPositionCheck.java [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+ * invalid by this check.
+ * </p>
+ * <p>
+ * To configure the check:
+ * </p>
+ * <pre>
+ * &lt;module name="InvalidJavadocPosition"/&gt;
+ * </pre>
+ * <p>
+ * The following code produces a violation because Javadocs should be before all annotations of
```
[the remark this answers is not in the document]
Comment (rnveach) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/InvalidJavadocPositionCheck.java]: I used the words before because the javadoc can end on the same line as the annotation, making above wrong.
Example: `/** Test */ @Annotation public void method() {}`

File: src/xdocs/checks.xml (hunk 4 above)
Comment (rnveach) [reply src/xdocs/checks.xml]: I don't see how, `correct` is just the negative of `invalid`.

File: src/it/java/com/google/checkstyle/test/chapter7javadoc/rule734nonrequiredjavadoc/InvalidJavadocPositionTest.java [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+    }
+
+    @Test
+    public void testDefault() throws Exception {
+        final String msg = getCheckMessage(InvalidJavadocPositionCheck.class,
+            "invalid.position");
+
+        final String[] expected = {
+            "1:9: " + msg,
+            "3:1: " + msg,
```
[the remark this answers is not in the document]
Comment (rnveach) [reply src/it/java/com/google/checkstyle/test/chapter7javadoc/rule734nonrequiredjavadoc/InvalidJavadocPositionTest.java]: Done.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/InvalidJavadocPositionCheck.java [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+ * /**
+ *  * This comment looks like javadoc but it at an invalid location.
+ *  * Therefore, the text will not get into TestClass.html and the check will produce a violation.
+ *  *&#47;
+ * public class TestClass {
+ * }
+ * </pre>
+ * @since 8.22
+ */
+@FileStatefulCheck
```
[the remark this answers is not in the document]
Comment (rnveach) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/InvalidJavadocPositionCheck.java]: Done.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/InvalidJavadocPositionCheck.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/InvalidJavadocPositionCheck.java]: ok

File: src/xdocs/checks.xml (hunk 4 above)
Comment (romani) [reply src/xdocs/checks.xml]: it was a one more hint that something wrong with name.
let skip this, as we do not have better name, lets use what we have.

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/invalidjavadocposition/InputInvalidJavadocPosition.java [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+/** violation */
+import java.lang.String;
+
+/** violation */
+/** valid */
+class InputInvalidJavadocPosition {
+    /** violation */    
+}
+/* ignore */
+class InputInvalidJavadocPosition2 {
```
Comment (romani) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/invalidjavadocposition/InputInvalidJavadocPosition.java]: please add testscase:
```
/** valid */
/* ignore */
class InputInvalidJavadocPosition2 {

/** violation */
/** valid */
class InputInvalidJavadocPosition2 {
```

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/invalidjavadocposition/InputInvalidJavadocPosition.java [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+    void /** violation */ method20() {}
+    void method21 /** violation */ () {}
+    void method22(/** violation */) {}
+    void method23() /** violation */ {}
+    void method24() { /** violation */ }
+    void method25() { /** violation */ int variable; }
+}
+@Deprecated
+/** violation */
+class InputInvalidJavadocPosition3 {}
```
Comment (romani) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/invalidjavadocposition/InputInvalidJavadocPosition.java]: please  add test case:
```
/** vaild */
@Deprecated
class InputInvalidJavadocPosition3 {}
```

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/invalidjavadocposition/InputInvalidJavadocPosition.java (hunk 8 above)
Comment (rnveach) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/invalidjavadocposition/InputInvalidJavadocPosition.java]: Second item is already there above `InputInvalidJavadocPosition`.
First item was added.

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/invalidjavadocposition/InputInvalidJavadocPosition.java (hunk 9 above)
Comment (rnveach) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/invalidjavadocposition/InputInvalidJavadocPosition.java]: Done.

**Comment by rnveach** [comment]
> Please add to ut tests cases:
http://rveach.no-ip.org/checkstyle/regression/reports/247/Orekit/xref/src/main/java/org/orekit/propagation/semianalytical/dsst/utilities/CoefficientsFactory.java.html#L39

Done.
